### ISSUE TITLE
What is/are the allegedly massive problem(s) in Celluloid?

### ISSUE THREAD

**Issue by digitalextremist** [celluloid/culture#8]
@e2 your emails are sitting in my inbox and I just can't get to them. I am sorry, but I want to know what you are saying. I have tried for more than 8 days, but I can't see a time when I will be able to right now:

* Email one: 4,860 words.
    * That's just your latest reply.
    * The thread in its entirety is 15,228 words.
    * I anticipate if I replied, the reply to back will be 10,000 words.
* Email two: 1,271 words.
    * Within 50 minutes.

Dude, no. This is not OK levels of email. Seriously man.

But the issue as I said isn't the sheer enormity of the emails. It's not *primarily* time investment. It's that I start my days from a positive place of building up momentum, even *determination* - and I want to keep it there and end it there.... but your words are just this continual refrain of how awful `Celluloid` is. It's like a receipt for sure failure to read those emails, and I am filtering out whatever isn't going to at least have a *chance* at success.

Maybe somewhere in there it gets to be something I can do something with. I didn't make it there each time I tried. My feeling is, if `Celluloid` is *that bad* and it's not just hyperbole ( as has been a doubt raised enough times for it to be really worrisome to me if I am going to spend time reading something that might just end up being this grandiose pseudo-non-fiction narrative ) me reading about how bad it is is not going to fix it. You have me *curious* though; from what *I* see, `Celluloid` is a lot of awesome things, and headed for a lot more. It's extremely poignant and viable. But us hearing what the specific failings are once, clearly, without repetition and just a straight-up beat down of the core team's competence in ever paragraph, will get the problems fixed *if they exist* -- and it can be even better.

Up to now we've each/all shown extreme respect for your opinions, but for example celluloid/celluloid#563 your cogence is just sort of 50/50, maybe more like 20/80 in dialog. And because this has become such a huge recurring situation and I really want there to be not just peace but clarity and momentum, I am going to very politely ask for you to just take the gloves off about what you see, but just *say it once, say it clearly, be prepared to defend your viewpoint* and **move on** to the next point. I am sorry I could not reply by email to let you know I cannot reply to you email, I am terrified to reply to your emails right now. It is this huge psychological burden to even think I have that large of an email in my otherwise perfectly triaged and organized and thoroughly read inbox.

So I open the floor to @e2 to christen the `diatribe` label in `celluloid/culture` now. Please @e2 disassemble your email(s). If there is anything in them that pertains only to me, please resend me only that by email. I want to read that, I care. Otherwise post *terse* and *succinct* issues here. If you need to splinter out into several sub-issues and reference this one and make a monumentally large network of complaints: *have at it*.

Otherwise I am just going to mark those emails read and really hope you take me up on this invitation.

**Comment by tarcieri** [comment]
@e2 I would have to second what @digitalextremist says. It's hard to separate your actual concerns from invective, hyperbole, and what is otherwise utter nonsense.

Let's take just the very first line of your complaints from #563:

> it's waaaaaay too complex (metaprograming metaprograming - yes that's correct)

That's not correct. There is in fact no metaprogramming going on whatsoever. There are two things going on, neither of which has to do with metaprogramming:

1) Celluloid maps Ruby dispatch onto the Celluloid actor protocol. This isn't "metaprogramming". Alan Kay would describe it as "late binding". Not metaprogramming
2) Celluloid was using reflection to determine if a particular "message" (in both the OOP and actor sense) is valid for a particular object/actor. Again, not metaprogramming

And this is line one of a huge diatribe that is but one of many diatribes in one of many diatribe-filled threads.

tl;dr: You're really hard to work with. You're doing great work, but if this were a company and I were your boss I would fire you.

**Comment by e2** [comment]
@tarcieri - that's why I prefer to do things case-by-case. I couldn't come up with a single specific case to justify the code (justifying use cases is different than justifying code).

@digitalextremist generalizes all the time (which just derails discussions for me, because I try to respond to the generalities). "We need to be sensitive" is not a specific, constructive argument.

> There is in fact no metaprogramming going on whatsoever. 

If someone sets up a pool with workers with 0.16, and their worker method has a crashing `#inspect` ... and they call a method with bad parameters .... then understanding what's going wrong is tricky to say the least. I won't even mention writing and debugging specs.

A lot of people are going to want to use Celluloid without the deep understanding. That's why I said "way too complex".

> There is no way to respond to you, because your arguments are ever-expanding gish gallops

Noted. I asked for objections and expected clear, specific, code-related use-cases (such as I'd get on many other open source project, like JRuby or RSpec). I got just general "speculations" as to what the code may be important for. Responding to them was my mistake.

> tl;dr: You're really hard to work with.

It goes both ways. I privately mentioned to @digitalextremist that I want to back off from further work in Celluloid. I wasn't the one not listening this time...

**Comment by digitalextremist** [comment]
>  "We need to be sensitive" is not a specific, constructive argument.

That's an axiom/premise, talking about foundations for conduct. It's a fundamental, not a generality.

> generalizes all the time

...is a generality.

@e2 25-33% of your comments to #563 were after you "backed off" which is half of why I posted this. 

The other half of why is because what you said by email ( and the continued stance you take on issues ) is authoritative and disapproving or pointed and expectant and disappointed in tone... it signals that there is something lacking or wrong, which I naturally try to address. Because I respected your opinion, in this thread I asked for your specific input on why you have that perspective. I wasn't airing you out other than posting specific word-counts to illustrate the level of workload you backing off looks like. I was not drawing you *back* in, I was asking you to bring out the things still going on in your mind, i.e:

> Noted. I asked for...

Therefore not noted, going right into another series of stabs immediately, because it's a living issue to you. You missing the point replied to, and the point I've tried to make to you since the moment we first exchanged words: to say "hey hold on" is not to delay you, it is to say "hey hold on" ... We don't have sponsored maintainers to respond and react with research and code in your sense of what's timely, before you take it as a personal affront or lack of service to the community, or both, or more. All it meant was "hey hold on" ... "please wait" ... but then the thread kept growing, compounding the "hey hold on"

---

It sounds like you're resolved in either dropping or privately holding your sentiments. I'll hope then that they don't get sprinkled over threads headed toward simple, clear, progress in the way we are working at this moment. To a lot of us, staying on top of *just the dialog* is extremely important, as important or more important than quickly making code changes. So when you post, it causes one or more of us to stand at attention and come to the thread.

**Comment by e2** [comment]
@digitalextremist - we're incompatible for such a project. You overgeneralize, while I expect PRs and code comments to do the talking. I'm the wrong guy for such a project. As for the emails - reading takes less time than writing (for most people), and I pretty much covered *everything* "once and for all". If you want to say something without me replying - just say so.

I have a ton of bad experiences with Celluloid - so I'm going to be biased. If people had such issues with projects I was involved with, I'd appreciate the feedback (not that I'd feel good about it at first - I'm human too). Feedback is always better than no feedback. I understand what it's like to be a "Celluloid beginner". I also get that Celluloid isn't how it is because of "sloppyness".

I didn't mind getting my hands dirty fixing what I'd otherwise silently complain about - but this resulted in too much non-technical friction (the opposite of what I expected from a project so deep into technicalities like Celluloid).

I wanted Celluloid to take into account the "casual user's" perspective - a lot of what I struggled with would've been handled by additional Wiki entries. Ideally though, I wanted to prevent issues in the code. Now I'm spending time neither writing code nor Wiki entries.

@digitalextremist - you're taking stuff personally. It's not that "Celluloid is bad" (which is a generalization), it's just that "issue X is a pain in the ass". If you take things like that personally, the arguments become non-technical (and thus, a waste of time). If someone is expecting a stricly technical response - try to hold back from going into generalizations (e.g. "we need to be sensitive about changes"). Meanwhile, I'll learn to hold back from addressing such responses.

@tarcieri - for most issues, I want to get to the core and resolve what's going on. When the issue stops being strictly technical, that's when "I'm hard to work with" interferes with most. I did a ton of debugging with the current codebase - and removing a bit of code for the sake of easier debugging means a lot to me. "Production-related" arguments or "archeology" should be specific enough to deal with - or suddently the "total time spent" trying to get a debugging improvement costs more than the benefit.

My core "complaints" (I'm just listing them FWIW):
- active + stable (in terms of tests) development doesn't happen on master
- it's unpredictable how a PR will be handled
- @digitalextremist goes into generalizations (which doesn't work well if someone like me wants to just get stuff done/fixed or a simply "yes"/"no" reply)
- issues like tests/debugging/tracing don't really seem to matter to anyone (they matter, but they are not a priority)
- too much focus on performance and production-compatibility
- lots of class renaming on non-master "release branches"
- personal projects dictate what goes in and what doesn't (instead of specs or at least examples)

Overall I'm just discouraged - and I'm trying to express and understand my attitude (learn from it), rather than change it and "pretend everything is ok".

**Comment by tarcieri** [comment]
@e2 gonna reiterate this...

> Worse, you don't seem to listen. When I do try to post responses, you seem to simply ignore them.

Are you even listening? Nothing you've said seems to have anything to do with anything I've said. It feels like a total communications breakdown.

**Comment by e2** [comment]
@digitalextremist 

> That's an axiom/premise, talking about foundations for conduct. It's a fundamental, not a generality.

Still, it's not specific. What else can I call it? Obscure? Vague? Fuzzy? Ambigous?

> @e2 25-33% of your comments to #563 were after you "backed off" which is half of why I posted this.

I was just responding.

> Therefore not noted, going right into another series of stabs immediately

I'm just replying.

> "hey hold on"

I'm assuming that if someone is busy, they won't have time to reply. Or their reply will give me some hint on how to organize my own work.

**Comment by e2** [comment]
@tarcieri , @digitalextremist - ask me something specific. Or just tell me specifically what behavior you expect from me.

**Comment by digitalextremist** [comment]
> Please @e2 disassemble your email(s). If there is anything in them that pertains only to me, please resend me only that by email. I want to read that. I care. Otherwise post terse and succinct issues here. If you need to splinter out into several sub-issues and reference this one and make a monumentally large network of complaints: have at it.

---

> ### What is/are the allegedly massive problem(s) in Celluloid?

---

These questions are designed for you to repeat yourself "once and for all" so you are fully heard. The previous attempts did not work... but if you notice, you are referenced in more and more issues and PRs because your perspective is obviously and unquestionably valid somehow. I started this thread to let you do exactly what you wanted to do: voice the "frustrated user" ... despite the anguish all around. That's really important, and it's not taken personally, just taken seriously.

Please take some time to use this space well, and give time for replies and discussion. It can go well? @e2, please take my specific promise to heart: I was not singling you out, violating our private conversational boundaries, etc. I've indicated emails exist, and cited word counts. What you've been insisting has been intended to stand out, so I'm pointing it out and saying "hey what was all that?"

Reading thousands of words is work, no matter how fast in comparison to "letting code talk" ... please keep it terse, or just pass on this for now. `What is/are the allegedly massive problem(s) in Celluloid?`

**Comment by e2** [comment]
@digitalextremist - answered above (search for "FWIW")

**Comment by digitalextremist** [comment]
> @e2 core "complaints" ...

1. active + stable (in terms of tests) development doesn't happen on master
2. it's unpredictable how a PR will be handled
3. @digitalextremist goes into generalizations (which doesn't work well if someone like me wants to just get stuff done/fixed or a simply "yes"/"no" reply)
4. issues like tests/debugging/tracing don't really seem to matter to anyone (they matter, but they are not a priority)
5. too much focus on performance and production-compatibility
6. lots of class renaming on non-master "release branches"
7. personal projects dictate what goes in and what doesn't (instead of specs or at least examples)

---

This will be longer than I would like to be writing right now. Knowing you do not listen, I will post this, lock this for a while, and reopen at a random unannounced point.

---

# 1. Response

OK. This fits into the "time window" comment. "Does not happen, *yet*" I mean. That is the goal we're working toward, on the way to `1.0` but we've had a major change of structure, so we had to handle that in a large parallel "master" branch called `0.17.0-prerelease` ... which might continue for another few weeks, or couple months. To me it depends on how quickly we can standardize and clean up `0.17.0` in real world implementations. I know you don't care about real world implementations, that's Complaint 5. But just hear our priorities for a moment: we want mature releases, stable branches that we can support in implementations depending on us.

> If may change this complaint to a test spec style assertion, "active + stable master" is `pending` one or two more release branches ( which there have been exactly one of in the last 1-2 years )?
> **In this time we plan to take, we will:**
> * Execute `GSoC` operations.
> * Wash out the several years of multiple ways of doing things.
> * Our most involved maintainer becoming a beloved advisor at a distance.
> * Losing several core behaviors all over the place.
> * Enforcing object naming conventions across the Celluverse.
> * Booting up Celluloid Culture standards.
> * Switching to microgems.

Now that this is clearly specified, your refrain of "I don't care about releases" is moot, and if you continue saying that I will be in the market for a way to mute you across all repositories until the above happens. Which flows nicely into Complaint 2. By the way that also answers Complaint 6.

# 2. Response

PR's which obviously have nothing to do with the response to Complaint 1, and have no regard for our response to Complaint 5 will sit or be closed until the code is relevant to Complaint 1's real world situation, which means continual rebasing -- but not many object name changes and file location changes though, those are mostly done. No dialog will be conducted on them. Speaking for myself I wouldn't mind adding a `relevant` label and only engaging those labeled as such except in more leisurely moments. That gem situation was a *bitch* though, let me tell you. *Reaching 1.0 is critical.* Deviating from a course continually PR to PR and getting into dissertation length "conversations" is going to debilitate us. It's already caused extreme noise and is sending the exact wrong signal to the world around us. One individual ( you ) has created a perception of disorder, disarray, discord, etc, which I want no comment on from you.

`FACT = "@e2 is creating an environment discouraging dialog, but claiming this is an environment which discourages dialog, while not actually engaging in dialog."`

> In other words, you can stop talking +90% of the time. Will revisit that ratio in like 90 days.

# 3. Response

This one had no specific evidence, but that's OK. I will translate based on the other things you've said; you do not want me to deal with you as a psychological element in the community, and a moving object in our air-traffic controlling situation. My continual "psychological" priority of creating an environment where we give a very, very large shit about how people feel when they work together under the Celluloid name very much bothers you. Not that you don't want an environment of respect, but that you want the environment to be pragmatist only and entirely focused on work, which negates so much of everything else you've said. But more specifically ( talking about the fundamentals, so this will sound like a generality ) subtleties of human interaction do not seem to be clear to you, which I choose to believe rather than you being a deliberate violator of other people's good day.

A lot of people have creative ideas about how to do things, which does not mean we're guided by pet projects, it means we're not here to debate the minutia or to deliver books of words to the maintainers of jRuby, Rubinius, etc. We expect them to be otherwise occupied with things they actually enjoy coding, and partially engaged in backfacing fixes, like we do. We do this because we want to do this *and* we do this correctly. Those items are not always perfectly in sync, but they are in balance. See: Time Window comment. This is one of my favorite examples of what you will shake a fist at all day long, and I will gladly put on a shirt and wear. I searched for the exact quote from an interaction between @niamster and @tarcieri where @niamster said "it's not pretty!" And @tarcieri said essentially, "it just needs to work, then it can be pretty later." We still have a few ugly releases ahead, then we'll polish up and we'll be interested in the ways to become more and more perfect. Right now we want to be feature complete, and stable. And part of "stable" in our sphere is "performant" to a large degree.

But regrouping to the complaint itself: I will always evade you and not answer your demanding questions if you aren't aware of what I just said, and acting like that is important to you, because this is not about two-word conversations, or 20,000 word ones. It's about creative and mannerly interactions between people who spend a lot of their time together because we are not assholes to each other, but more than that, because we've grown accustomed to how each other works and thinks, and you've so far type-cast yourself as K.I.T.T. from Knight Rider. Early on when he is pissed off at everything Michael Knight does because it doesn't make rational sense, except that he drives fast, quells crime, and without fail gets the girl.

# 4. Response

Untrue. This is that time window thing again. Try walking into someone's bathroom while they're taking a shower and being like "you really don't like crystal clear mirrors you barbarians" and see how relevant that sounds coming out of your mouth. That's about how relevant that comment was. The tests fail, and we aren't surprised, but that's not because we don't care, it is because most of us have 1-3 full time jobs and are not supposed to be hating ourselves and being all pissed off at Celluloid for it being in the shower right now. Let it finish showering.

# 5. Response

I think I will dedicate another comment to this one by itself later, but so far you should already the gist that there are three elements at work: performance, production maturity, and perfection, and we strive for a balance, but like I said, there is the time window in place, and look at the code all you want -- until you understand our mental frame ( not our mindset ) you will be fooled into believing what you said. We just need to focus on those because we're not writing some boutique clique or an invisibly one-dimensional or uncomplicated scope. We're working in a very hotly contended for space, and that's fine. We don't need to share your priorities, but if you don't share ours, you can just try to persuade, not be deaf to us and play like you're our friend and in reality be secretly hating how we are, trying to change how we are, or what we have, because of all those days and nights you've been frustrated by us. That's the backdrop for a spy novel or saboteur persona. In that case your cloaking as someone well meaning would be really confusing, and the degree to which you pay no attention to what we say leads me to believe more and more that you have no interest at all in Celluloid, or in our community of coders, but purely to check a box on your anti-bucket list, of things annoying you.

# 7. Response

Without listening to our priorities and expectations for our own future, you are not qualified to distinguish between pet projects and divide-and-conquer distribution of large issues that are better off being taken off the beaten path and worked on without disturbance. Right now, it is true that a lot of documentation, examples, and test cases are not present, but that is not because we're off doing irrelevant things or ignoring responsible tasks, we just don't subscribe to your idea of what should be done, but you only see your way. Complaint 7 is sort of the tip of the iceberg where you were starting to actually answer my original question honestly rather than what you ended up saying which I just responded to.

---

I am going to freeze this here for a while, but in the mid-term future we can continue if you are interested. Please remember, for at least 90 days you are well advised to reduce your verbal output by around 90% because it might take that much just for you to have an opportunity to listen, *understand* and then think about what other people say... and your demeanor is not fitting for the creative and vibrant environment we're accustomed to, no matter how dormant things have been from time to time. This is not muzzling you, it's inviting you to have fun because if you 90% shut up you will have more fun, promise. You are welcome to submit any PR's you want, or to fade out for a while, or whatever you want to do *EXCEPT* be a demanding and condescending asshole. You have skills, you have endurance, these are awesome; but they come at far too high a price. I hope you stick around, but I hope you change your ways. Thank you.

**Comment by tarcieri** [comment]
> ask me something specific. Or just tell me specifically what behavior you expect from me.

I already gave you specific criticisms, and you didn't seem to read them or care. I can't tell.

The criticism I gave you was that you don't listen or just generally seem to ignore what I say. With ^^^ message we've reached inception level: I criticized you for not listening, then you responded I didn't give specific criticisms, in the process not listening to my criticism that you aren't listening.

So, listen please?
